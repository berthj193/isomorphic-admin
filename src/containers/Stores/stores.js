import React, { Component } from 'react';
import PropTypes from 'prop-types';
import autoBind from 'react-autobind';
import columns from './columns';
import storeQueries from '../../graphql/stores';
import withStoresManager from '../../components/apolloProviders/withStoresManager';
import ItemList from '../../components/list/item-list';

import Details from './details';
import DeleteButton from '../../components/uielements/deleteButton';

export class Stores extends Component {
  static propTypes = {
    stores: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        warehouse: PropTypes.shape({ name: PropTypes.string }),
        company: PropTypes.shape({ name: PropTypes.string }),
        channel: PropTypes.shape({ name: PropTypes.string }),
        store_products: PropTypes.arrayOf(PropTypes.object),
        created_at: PropTypes.string,
      }),
    ),
    error: PropTypes.string,
    mutate: PropTypes.func.isRequired,
  };

  static defaultProps = {
    stores: [],
  };


  constructor() {
    super();
    autoBind(this);
  }

  dataSource() {
    return this.props.stores.map(({
      id, key, created_at, store_products = [],
      warehouse = {}, company = {}, channel = {},
    }) => ({
      key,
      id,
      warehouse_name: warehouse.name,
      company_name: company.name,
      channel_name: channel.name,
      created_at,
      products_count: store_products.length,
    }));
  }

  detailsPath = ({ pathname, record: { key } }) =>
    `${pathname}/${key}`;

  deleteStore = id => e => {
    e.preventDefault();
    e.stopPropagation();
    this.props.mutate({
      mutation: storeQueries.destoryStore,
      variables: {
        id,
      },
      refetchQueries: [
        {
          query: storeQueries.getList,
        },
      ],
      messages: ['Store has been removed successfully'],
    });
  };

  actionColumn = ({ id }) => (
    <ul className="table-action">
      <li>
        <DeleteButton
          tooltipText="Remove a store"
          onConfirm={this.deleteStore(id)}
        />
      </li>
    </ul>
  );

  columns = [
    ...columns,
    {
      title: 'Action',
      dataIndex: '',
      key: 'a',
      render: this.actionColumn,
    },
  ];

  render() {
    return (
      <ItemList
        data={this.dataSource()}
        error={this.props.error}
        title="Stores"
        columns={this.columns}
        detailsComponent={Details}
        buttons={[{
          title: 'New form',
          link: '/dashboard/stores/new',
        }]}
        detailsPath={this.detailsPath}
      />
    );
  }
}

export default withStoresManager(Stores,
  {
    query: storeQueries.getList,
  }
);
