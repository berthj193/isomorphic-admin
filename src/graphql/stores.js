import gql from 'graphql-tag';

const storeReturnData = `
    id
    key
    options
    created_at
    channel {
      id
      name
    }
    warehouse {
      id
      name
    }
    legal_entity {
      id
      name
    }
`;

export const storesAutocompleteQuery = `
query {
  stores {
    id
    key
  }
}`;

export const storeListQuery = `
query {
  stores {
    ${storeReturnData}
  }
}`;

export const storeDetailsQuery = `
query stores (
  $store_key: String!
) {
  stores(
    store_key: $store_key
  ) {
    ${storeReturnData}
  }
}`;

export const createStoreQuery = `
  mutation createStore(
    $key: String!
    $options: JsonType!
    $legal_entity_id: ID!
    $warehouse_id: ID!
    $channel_id: ID!
  ) {
    createStore(
      key: $key
      options: $options
      legal_entity_id: $legal_entity_id
      warehouse_id: $warehouse_id
      channel_id: $channel_id
    )
    {
      ${storeReturnData}
    }
  }
`;

export const updateStoreQuery = `
  mutation updateStore(
    $id: ID!
    $key: String!
    $options: JsonType!
    $legal_entity_id: ID!
    $warehouse_id: ID!
    $channel_id: ID!
  ) {
    updateStore(
      id: $id
      key: $key
      options: $options
      legal_entity_id: $legal_entity_id
      warehouse_id: $warehouse_id
      channel_id: $channel_id
    )
    {
      ${storeReturnData}
    }
  }
`;

export const destroyStoreQuery = `
mutation destroyStore (
  $id: ID!
) {
  destroyStore (
    id: $id
  )
}`;

export default {
  getDetails: gql`${storeDetailsQuery}`,
  getList: gql`${storeListQuery}`,
  createStore: gql`${createStoreQuery}`,
  updateStore: gql`${updateStoreQuery}`,
  destoryStore: gql`${destroyStoreQuery}`,
  storesAutocomplete: gql`${storesAutocompleteQuery}`,
};
