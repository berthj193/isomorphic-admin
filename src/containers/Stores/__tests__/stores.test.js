import React from 'react';
import { Stores } from '../stores';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('<Stores />', () => {
  const defaultProps = {};

  const getComponent = (props = { mutate: () => null }) => shallow(<Stores {...Object.assign(defaultProps, props)} />);

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render properly', () => {
    expect(toJson(getComponent())).toMatchSnapshot();
  });

  it('should map stores', () => {
    const mockStores = [
      {
        id: '123',
        warehouse: { name: 'My Little Warehouse' },
        company: { name: 'Shmetterling' },
        channel: { name: 'CNN' },
        created_at: new Date(1990).toISOString(),
        store_products: [{}],
      },
      {
        id: '133',
        warehouse: { name: 'My Bigger Warehouse' },
        company: { name: 'Kummerspeck' },
        channel: { name: 'BBC' },
        created_at: new Date(1992).toISOString(),
        store_products: [{}, {}, {}],
      },
    ];
    const component = getComponent({ stores: mockStores });

    expect(component.instance().dataSource()).toMatchSnapshot();
  });
});

