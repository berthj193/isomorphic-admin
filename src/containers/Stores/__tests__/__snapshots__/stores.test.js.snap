// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stores /> should map stores 1`] = `
Array [
  Object {
    "channel_name": "CNN",
    "company_name": "Shmetterling",
    "created_at": "1970-01-01T00:00:01.990Z",
    "id": "123",
    "key": undefined,
    "products_count": 1,
    "warehouse_name": "My Little Warehouse",
  },
  Object {
    "channel_name": "BBC",
    "company_name": "Kummerspeck",
    "created_at": "1970-01-01T00:00:01.992Z",
    "id": "133",
    "key": undefined,
    "products_count": 3,
    "warehouse_name": "My Bigger Warehouse",
  },
]
`;

exports[`<Stores /> should render properly 1`] = `
<withRouter(ItemList)
  buttons={
    Array [
      Object {
        "link": "/dashboard/stores/new",
        "title": "New form",
      },
    ]
  }
  columns={
    Array [
      Object {
        "dataIndex": "id",
        "key": "id",
        "title": "ID",
      },
      Object {
        "dataIndex": "key",
        "key": "key",
        "title": "Key",
      },
      Object {
        "dataIndex": "warehouse_name",
        "key": "warehouse_name",
        "title": "Warehouse",
      },
      Object {
        "dataIndex": "created_at",
        "key": "created_at",
        "title": "Created at",
      },
      Object {
        "dataIndex": "channel_name",
        "key": "channel_name",
        "title": "Channel",
      },
      Object {
        "dataIndex": "products_count",
        "key": "products_count",
        "title": "Products Count",
      },
      Object {
        "dataIndex": "",
        "key": "a",
        "render": [Function],
        "title": "Action",
      },
    ]
  }
  data={Array []}
  detailsComponent={[Function]}
  detailsPath={[Function]}
  title="Stores"
/>
`;
